### PR TITLE
enc : add support for wrap mode

### PR DESCRIPTION
**Pull request by EasySec**
Wrap modes are not explicitly UNsupported (as for XTS or AEAD modes), but still not working (unless I am mistaken).
That's why I believe there no need to update the documentation.
Tested with different values of -bufsize arg.

### CODE REVIEW

**Comment by t8m** [comment]
I suppose documentation how the enc command works with the wrap mode ciphers would need to be added and also at least some test of the command with a wrap mode cipher.

**Comment by EasySec** [comment]
OK, I will try some documentation. Could you please give any guidance / example I could look at to add 'command line' tests ?

**Comment by EasySec** [comment]
#17702

**Comment by EasySec** [comment]
What about the doc changes ?
Unfortunately I am not skilled enough in perl to add usefull tests cases.
:-(
remark : \*-wrap-\* modes where listed in the output of supported ciphers of the enc command (-list). 
Can we consider it is not really a new feature ?

File: doc/man1/openssl-enc.pod.in [hunk 1]
```diff
@@ -274,14 +274,22 @@ able to roll back upon authentication failure.  The AEAD modes currently in
...
-exposing AEAD modes is too great to allow.  These key/iv/nonce
+exposing AEAD modes is too great to allow. These key/iv/nonce
 management issues also affect other modes currently exposed in this command,
 but the failure modes are less extreme in these cases, and the
 functionality cannot be removed with a stable release branch.
 For bulk encryption of data, whether using authenticated encryption
 modes or other modes, L<openssl-cms(1)> is recommended, as it provides a
 standard data format and performs the needed key/iv/nonce management.
 
+key wrapping : input data cannot be streamed, so must must be processed
```
Comment (t8m) [inline doc/man1/openssl-enc.pod.in:285]: I'd write something like this: 
```
=head2 Key wrapping modes

When enc is used with key wrapping modes the input data cannot be streamed, meaning
it must be processed in a single pass...
```

File: doc/man1/openssl-enc.pod.in [hunk 2]
```diff
@@ -274,14 +274,22 @@ able to roll back upon authentication failure.  The AEAD modes currently in
...
+in one single pass. Consequently, the input data size must be lower than
+the buffer size (-bufsize arg, default to 8*1024 bytes).
+The '*-wrap' ciphers require the input to be a multiple of 8 bytes long, 
+because no padding is involved.
+The '*-wrap-pad' ciphers allows any input length.
```
Comment (t8m) [inline doc/man1/openssl-enc.pod.in:290]: allows -> allow

**Review by t8m — CHANGES_REQUESTED** [review]

File: doc/man1/openssl-enc.pod.in (hunk 2 above)
Comment (EasySec) [reply doc/man1/openssl-enc.pod.in:290]: Thanks, I will update this commit.

**Review by t8m — APPROVED** [review]

File: doc/man1/openssl-enc.pod.in [hunk 3]
```diff
@@ -274,14 +274,23 @@ able to roll back upon authentication failure.  The AEAD modes currently in
...
 management issues also affect other modes currently exposed in this command,
 but the failure modes are less extreme in these cases, and the
 functionality cannot be removed with a stable release branch.
 For bulk encryption of data, whether using authenticated encryption
 modes or other modes, L<openssl-cms(1)> is recommended, as it provides a
 standard data format and performs the needed key/iv/nonce management.
 
+When enc is used with key wrapping modes the input data cannot be streamed, 
+meaning it must be processed in a single pass.
+Consequently, the input data size must be lower than
```
Comment (paulidale) [inline doc/man1/openssl-enc.pod.in:287]: _less than_ instead of _lower than_

File: doc/man1/openssl-enc.pod.in [hunk 4]
```diff
@@ -417,6 +429,12 @@ Base64 decode a file then decrypt it using a password supplied in a file:
  openssl enc -aes-256-ctr -pbkdf2 -d -a -in file.aes256 -out file.txt \
     -pass file:<passfile>
 
+AES key wrapping:
+
+ openssl enc -e -a -id-aes128-wrap-pad -K 000102030405060708090A0B0C0D0E0F -in file.bin
+ or
```
Comment (paulidale) [inline doc/man1/openssl-enc.pod.in:435]: Should the _or_ not have a space at the front?

**Review by paulidale — APPROVED** [review]
Looks good.  With unstreamable ciphers added, we could consider unstreamable AEAD schemes since the nasty failure case disappears.

Just one wording change.

File: doc/man1/openssl-enc.pod.in (hunk 3 above)
Comment (EasySec) [reply doc/man1/openssl-enc.pod.in:287]: Thanks again ! As you may have noticed, English is not my mother tongue.
;-)

File: doc/man1/openssl-enc.pod.in (hunk 4 above)
Comment (EasySec) [reply doc/man1/openssl-enc.pod.in:435]: I will add a space.

Comment (paulidale) [reply doc/man1/openssl-enc.pod.in]: I meant drop the leading spaces so it's not a mono-space font.

**Comment by paulidale** [comment]
Waiting on the dropping of two spaces.

**Comment by paulidale** [comment]
Merged after cleaning end of line whitespace.  Thanks for the contribution.

**Pull request closed without merging**